File: packages/firebase_app_check/firebase_app_check/ios/firebase_app_check/Sources/firebase_app_check/FLTAppCheckProvider.m
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FLTAppCheckProvider.h"

@implementation FLTAppCheckProvider

- (id)initWithApp:app {
  self = [super init];
  if (self) {
    self.app = app;
  }
  return self;
}

- (void)configure:(FIRApp *)app
     providerName:(NSString *)providerName
       debugToken:(NSString *)debugToken {
  if ([providerName isEqualToString:@"debug"]) {
    if (debugToken != nil) {
      // We have a debug token, so just need to stuff it in the environment and it will hook up
      char *key = "FIRAAppCheckDebugToken", *value = (char *)[debugToken UTF8String];
      int overwrite = 1;
      setenv(key, value, overwrite);
    }
    FIRAppCheckDebugProvider *provider = [[FIRAppCheckDebugProvider alloc] initWithApp:app];
    if (debugToken == nil) NSLog(@"Firebase App Check Debug Token: %@", [provider localDebugToken]);
    self.delegateProvider = provider;
  }

  if ([providerName isEqualToString:@"deviceCheck"]) {
    self.delegateProvider = [[FIRDeviceCheckProvider alloc] initWithApp:app];
  }

  if ([providerName isEqualToString:@"appAttest"]) {
    if (@available(iOS 14.0, macCatalyst 14.0, tvOS 15.0, watchOS 9.0, *)) {
      self.delegateProvider = [[FIRAppAttestProvider alloc] initWithApp:app];
    } else {
      // This is not a valid environment, setup debug provider.
      self.delegateProvider = [[FIRAppCheckDebugProvider alloc] initWithApp:app];
    }
  }

  if ([providerName isEqualToString:@"appAttestWithDeviceCheckFallback"]) {
    if (@available(iOS 14.0, *)) {
      self.delegateProvider = [[FIRAppAttestProvider alloc] initWithApp:app];
    } else {
      self.delegateProvider = [[FIRDeviceCheckProvider alloc] initWithApp:app];
    }
  }
}

- (void)getTokenWithCompletion:(nonnull void (^)(FIRAppCheckToken *_Nullable,
                                                 NSError *_Nullable))handler {
  // Proxying to delegateProvider
  [self.delegateProvider getTokenWithCompletion:handler];
}

@end

File: packages/firebase_app_check/firebase_app_check/ios/firebase_app_check/Sources/firebase_app_check/FLTAppCheckProvider.swift
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation
import FirebaseAppCheck
import FirebaseCore

class FLTAppCheckProvider: NSObject, AppCheckProvider {
    var app: FirebaseApp?
    var delegateProvider: AppCheckProvider?
    
    init(app: FirebaseApp) {
        super.init()
        self.app = app
    }
    
    func configure(app: FirebaseApp, providerName: String) {
        switch providerName {
        case "debug":
            let provider: AppCheckDebugProvider? = AppCheckDebugProvider(app: app)
            if let debugToken = provider?.localDebugToken() {
                print("Firebase App Check Debug Token: \(debugToken)")
            }
            self.delegateProvider = provider
            
        case "deviceCheck":
            self.delegateProvider = DeviceCheckProvider(app: app)
            
        case "appAttest":
            if #available(iOS 14.0, macCatalyst 14.0, tvOS 15.0, watchOS 9.0, *) {
                self.delegateProvider = AppAttestProvider(app: app)
            } else {
                // This is not a valid environment, setup debug provider.
                self.delegateProvider = AppCheckDebugProvider(app: app)
            }
            
        case "appAttestWithDeviceCheckFallback":
            if #available(iOS 14.0, *) {
                self.delegateProvider = AppAttestProvider(app: app)
            } else {
                self.delegateProvider = DeviceCheckProvider(app: app)
            }
            
        default:
            break
        }
    }
    
    func getToken(completion: @escaping (AppCheckToken?, Error?) -> Void) {
        // Proxying to delegateProvider
        delegateProvider?.getToken(completion: completion)
    }
} 